File: languages/java/solutions/src/main/java/dev/ruben/algorithms/fizzbuzz/FizzBuzz.java
package dev.ruben.algorithms.fizzbuzz;

import java.util.ArrayList;
import java.util.List;

public final class FizzBuzz {

    private FizzBuzz() {
    }

    /**
     * Returns the FizzBuzz representation of every integer from 1 through n.
     *
     * <p>Time complexity: O(n). Space complexity: O(n) for the returned list.</p>
     */
    public static List<String> generate(int n) {
        var result = new ArrayList<String>(n);

        for (int value = 1; value <= n; value++) {
            if (value % 3 == 0 && value % 5 == 0) {
                result.add("FizzBuzz");
            } else if (value % 3 == 0) {
                result.add("Fizz");
            } else if (value % 5 == 0) {
                result.add("Buzz");
            } else {
                result.add(Integer.toString(value));
            }
        }

        return result;
    }
}

File: languages/typescript/solutions/fizzbuzz/fizzbuzz.ts
/**
 * Returns the FizzBuzz representation of every integer from 1 through n.
 *
 * Time: O(n)
 * Space: O(n) for the returned collection
 */
export function fizzBuzz(n: number): string[] {
  const result: string[] = [];

  for (let value = 1; value <= n; value++) {
    if (value % 3 === 0 && value % 5 === 0) {
      result.push("FizzBuzz");
    } else if (value % 3 === 0) {
      result.push("Fizz");
    } else if (value % 5 === 0) {
      result.push("Buzz");
    } else {
      result.push(value.toString());
    }
  }

  return result;
}
